tools/utils/cancelation.js

```diff
@@ -0,0 +1,41 @@
+class CancelError extends Error {}
+
+export class CancelToken {
+  constructor(exec) {
+    let requested = false;
+    let message;
+
+    exec(function cancel(reason = "cancelation requested") {
```

Comment:
This pattern is similar to the function passed to the `Promise` constructor. The ability to cancel the token (like resolving or rejecting a `Promise`) remains cleanly separated from the token itself, which can only check whether cancelation has been requested.